src/mpid/common/shm/mpidu_shm_alloc.c

```diff
@@ -0,0 +1,664 @@
+
+        if (iter == 0) {
+            map_pointer = -1ULL;
+            goto fn_exit;
+        }
+    }
+
+  fn_exit:
+    return (void *) map_pointer;
+}
+
+typedef struct {
+    unsigned long long sz;
+    int loc;
+} ull_maxloc_t;
+
+/* Compute maxloc for unsigned long long type.
+ * If more than one max value exists, the loc with lower rank is returned. */
+static void ull_maxloc_op_func(void *invec, void *inoutvec, int *len, MPI_Datatype * datatype)
+{
+    ull_maxloc_t *inmaxloc = (ull_maxloc_t *) invec;
+    ull_maxloc_t *outmaxloc = (ull_maxloc_t *) inoutvec;
+    int i;
+    for (i = 0; i < *len; i++) {
+        if (inmaxloc->sz > outmaxloc->sz) {
+            outmaxloc->sz = inmaxloc->sz;
+            outmaxloc->loc = inmaxloc->loc;
+        } else if (inmaxloc->sz == outmaxloc->sz && inmaxloc->loc < outmaxloc->loc)
+            outmaxloc->loc = inmaxloc->loc;
+    }
+}
+
+/* Allreduce MAXLOC for unsigned size type by using user defined operator
+ * and derived datatype. We have to customize it because standard MAXLOC
+ * supports only pairtypes with signed {short, int, long}. We internally
+ * cast size_t to unsigned long long which is large enough to hold size type
+ * and matches an MPI basic datatype. */
+static int allreduce_maxloc(size_t mysz, int myloc, MPIR_Comm * comm, size_t * maxsz,
+                            int *maxsz_loc)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int blocks[2] = { 1, 1 };
+    MPI_Aint disps[2];
+    MPI_Datatype types[2], maxloc_type = MPI_DATATYPE_NULL;
+    MPI_Op maxloc_op = MPI_OP_NULL;
+    ull_maxloc_t maxloc, maxloc_result;
+    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+
+    types[0] = MPI_UNSIGNED_LONG_LONG;
+    types[1] = MPI_INT;
+    disps[0] = 0;
+    disps[1] = (MPI_Aint) ((uintptr_t) & maxloc.loc - (uintptr_t) & maxloc.sz);
+
+    mpi_errno = MPIR_Type_create_struct_impl(2, blocks, disps, types, &maxloc_type);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    mpi_errno = MPIR_Type_commit_impl(&maxloc_type);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    mpi_errno = MPIR_Op_create_impl(ull_maxloc_op_func, 0, &maxloc_op);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    maxloc.sz = (unsigned long long) mysz;
+    maxloc.loc = myloc;
+
+    mpi_errno = MPIR_Allreduce(&maxloc, &maxloc_result, 1, maxloc_type, maxloc_op, comm, &errflag);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    *maxsz_loc = maxloc_result.loc;
+    *maxsz = (size_t) maxloc_result.sz;
+
+  fn_exit:
+    if (maxloc_type != MPI_DATATYPE_NULL)
+        MPIR_Type_free_impl(&maxloc_type);
+    if (maxloc_op != MPI_OP_NULL)
+        MPIR_Op_free_impl(&maxloc_op);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+/* Try mapping shmem into processes address space symmetrically using address passed in
+ * through shm_seg->base_addr. Processes in different nodes coordinate after the function
+ * returns. */
+static int map_symm_shm(MPIR_Comm * shm_comm_ptr, MPIDU_shm_seg_t * shm_seg, int *map_result)
+{
+    int mpi_errno = MPI_SUCCESS, mpl_err = 0;
```

Comment:
Can you please check the comments for the original `MPIDIU_allocate_symshm_segment`? It seems that the logic "if it is MPIDIU_SYMSHM_OTHER_FAIL, it usually means no more shm segment can be allocated, thus the caller should stop retrying" is missing.

Besides, should we have error handling in this function when non-map failure happens?  Please see the original code.

Reply:
You are right! I must have introduce the error when I separated the allocation from the retrial code